subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/processors/MaxNParallelTestClassProcessor.java

```diff
@@ -76,9 +83,29 @@ public void processTestClass(TestClassRunInfo testClass) {
     @Override
     public void stop() {
         try {
-            CompositeStoppable.stoppable(processors).add(actors).add(resultProcessorActor).stop();
+            CompositeStoppable stoppable;
+            processorsLock.lock();
+            try {
+                stoppable = CompositeStoppable.stoppable(processors);
+                processors.clear();
+            } finally {
+                processorsLock.unlock();
+            }
+            stoppable.add(actors).add(resultProcessorActor).stop();
         } catch (DispatchException e) {
             throw UncheckedException.throwAsUncheckedException(e.getCause());
         }
     }
+
+    @Override
+    public void stopNow() {
```

Comment:
After more thought & local experiments, I realized that this method isn't going to work as anticipated because the `stop()` will likely have been invoked, clearing out the `processors` list & invalidating the associated `Actor`s.  I'm now thinking that this class needs to retain the "raw" `TestClassProcessor`s so that they can have `stopNow` called directly without going through the (potentially stopped) `Actor`.

Reply:
This should be better/fixed with the [next commit](https://github.com/gradle/gradle/pull/4190/commits/e7607102bec0ffadcd3770ef5133c0b0e4e07145)

Reply:
This is a case of "leaky abstraction" and will probably give us some more issues when we poke at this with integration tests. If this works out I'm okay with doing it like that as a first step, but we should probably revisit the actor infrastructure.